libcontainer/cgroups/cgroupv1/utils.go

```diff
@@ -0,0 +1,307 @@
+			continue
+		}
+		fields := strings.Split(txt, " ")
+		m := Mount{
+			Mountpoint: fields[4],
+			Root:       fields[3],
+		}
+		for _, opt := range strings.Split(fields[len(fields)-1], ",") {
+			seen, known := ss[opt]
+			if !known || (!all && seen) {
+				continue
+			}
+			ss[opt] = true
+			opt = strings.TrimPrefix(opt, CgroupNamePrefix)
+			m.Subsystems = append(m.Subsystems, opt)
+			numFound++
+		}
+		if len(m.Subsystems) > 0 || all {
+			res = append(res, m)
+		}
+	}
+	if err := scanner.Err(); err != nil {
+		return nil, err
+	}
+	return res, nil
+}
+
+// GetCgroupMounts returns the mounts for the cgroup subsystems.  Parameter
+// `all` indicates whether to return just the first instance or all the mounts.
+// This function is cgroupv1 specific and should not be used from cgroupv2 code.
+func GetCgroupMounts(all bool) ([]Mount, error) {
+	if v2() {
+		return nil, errUnified
+	}
+
+	f, err := os.Open("/proc/self/mountinfo")
+	if err != nil {
+		return nil, err
+	}
+	defer f.Close()
+
+	allSubsystems, err := ParseCgroupFile("/proc/self/cgroup")
+	if err != nil {
+		return nil, err
+	}
+
+	allMap := make(map[string]bool)
+	for s := range allSubsystems {
+		allMap[s] = false
+	}
+	return getCgroupMountsHelper(allMap, f, all)
+}
+
+// GetOwnCgroup returns the relative path to the cgroup docker is running in.
+func GetOwnCgroup(subsystem string) (string, error) {
+	if v2() {
+		return "", errUnified
+	}
+	cgroups, err := ParseCgroupFile("/proc/self/cgroup")
+	if err != nil {
+		return "", err
+	}
+
+	return getControllerPath(subsystem, cgroups)
+}
+
+func GetOwnCgroupPath(subsystem string) (string, error) {
+	cgroup, err := GetOwnCgroup(subsystem)
+	if err != nil {
+		return "", err
+	}
+
+	return getCgroupPathHelper(subsystem, cgroup)
+}
+
+func GetInitCgroup(subsystem string) (string, error) {
+	if v2() {
+		return "", errUnified
+	}
+	cgroups, err := ParseCgroupFile("/proc/1/cgroup")
+	if err != nil {
+		return "", err
+	}
+
+	return getControllerPath(subsystem, cgroups)
+}
+
+func GetInitCgroupPath(subsystem string) (string, error) {
+	cgroup, err := GetInitCgroup(subsystem)
+	if err != nil {
+		return "", err
+	}
+
+	return getCgroupPathHelper(subsystem, cgroup)
+}
+
+func getCgroupPathHelper(subsystem, cgroup string) (string, error) {
+	mnt, root, err := FindCgroupMountpointAndRoot("", subsystem)
+	if err != nil {
+		return "", err
+	}
+
+	// This is needed for nested containers, because in /proc/self/cgroup we
+	// see paths from host, which don't exist in container.
+	relCgroup, err := filepath.Rel(root, cgroup)
+	if err != nil {
+		return "", err
+	}
+
+	return filepath.Join(mnt, relCgroup), nil
+}
+
+// ParseCgroupFile parses the given cgroup file, typically from
+// /proc/<pid>/cgroup, into a map of subgroups to cgroup names.
+//
+// This function is cgroupv1-specific and must not be used from
+// cgroupv2 code.
+func ParseCgroupFile(path string) (map[string]string, error) {
```

Comment:
why is this function only for cgroup v1?  It can be used on cgroup v2 as well

Reply:
So, what it returns is a map of controllers to their paths, e.g.
```go
"pids" => "/user.slice/user-1000.slice"
"freezer" => "/"
"net_cls" => "/"
```
and so on. It is used to get the per-controller path. A caller has certain expectations (like the map keys being the controllers available).

For cgroupv2 (unified hierarchy), there is only one path, something like
```go
"" => "/user.slice/user-1000.slice"
```

so obviously a map is not needed here.

In fact we have cgroupv2-only implementation of parseCgroupFile here:
https://github.com/opencontainers/runc/blob/a891fee8c6fa7473893c5102b74922ac21f99415/libcontainer/cgroups/fs2/defaultpath.go#L75

but it's not public. If it is to be made public I think `GetOwnCgroup`/`GetInitCgroup` is a better API. If we want to have  `Get{Own,Init}Cgroup{,Path}` for v2, there are three ways:
1. have a separate v1 and v2 implementations (for v2, drop the `subsystem` argument)
2. have an implementation that works for both v1 and v2, ignoring the `subsystem` for v2 case.
3. have an implementation that works for both v1 and v2, checking the `subsystem` is available for v2 case.

I am very open to change all, just need to get more user cases. What use case do you have in mind?

Reply:
even without considering the hybrid model and just considering the v2 unified hierarchy with all the controllers assigned to cgroup v2, there could still be v1 named hierarchies, so a process  belongs to a cgroup v2 hierarchy and also to named hierarchies.  systemd for example creates a named hierarchy even on cgroup v2 when running a container that uses an old version of systemd.

If someone is interested, I think returning a map would still make sense on cgroup v2 to cover this case.

Kubernetes is currently using `ParseCgroupFile` to parse `/proc/$PID/cgroup` for another process, so it would be useful if it allows to specify the file path.

Reply:
Here 
https://github.com/kubernetes/kubernetes/blob/fb9e1946b0af89998b3e366d49399f6365c94c56/pkg/kubelet/cm/container_manager_linux.go#L887-L894 

As you can see, the value returned by `ParseCgroupFile` is used in different ways for v1 and v2 -- this is the primary reason I want to make it v1-specific. I mean, ideally the code above should look like

```go
func getContainer(pid int) (string, error) {
if cgroups.IsCgroup2UnifiedMode() {
	return cgroupv2.GetProcessCgroup(pid)
}

cgs, err := cgroupv1.ParseCgroupFile(fmt.Sprintf("/proc/%d/cgroup", pid))
if err != nil {
	return "", err
}
...
```

but since we don't have `cgroupv2.GetProcessCgroup()` (yet?), I can redo this patchset to not move `ParseCgroupFile()` (for now?)

Reply:
yes, I agree with you.  What do you think about `cgroupv2.ParseCgroupFile()` that is a wrapper around `cgroupv1.ParseCgroupFile()`?  The cgroupv1 version can still be used if someone cares about other hierarchies

Reply:
I have updated this PR to not move `ParseCgroupFile` to `cgroupv1` package.